abci/example/kvstore/kvstore.go

```diff
@@ -66,16 +67,35 @@ var _ types.Application = (*Application)(nil)
 type Application struct {
 	types.BaseApplication
 
+	// This implementation requires locking because only a single state is in use.
+	mtx          *tmsync.RWMutex
 	state        State
 	RetainBlocks int64 // blocks to retain after commit (via ResponseCommit.RetainHeight)
 }
 
 func NewApplication() *Application {
 	state := loadState(dbm.NewMemDB())
-	return &Application{state: state}
+
+	return &Application{mtx: new(tmsync.RWMutex), state: state}
+}
+
+func (app *Application) ReadLock() func() {
```

Comment:
I think it'd be better (in the "easier to read" sense) to just use the idiomatic/standard mutex calls rather than this kind of approach.